Tools/Algorithms/psatd.ipynb

```diff
@@ -0,0 +1,631 @@
+    "    fields_list.append(F)\n",
+    "if divB_cleaning:\n",
+    "    fields_list.append(G)\n",
+    "EBFG = zeros(dim, 1)\n",
+    "for i in range(EBFG.shape[0]):\n",
+    "    EBFG[i] = fields_list[i]\n",
+    "\n",
+    "# Define array of first-order time derivatives of all fields\n",
+    "fields_list = [dEx_dt, dEy_dt, dEz_dt, dBx_dt, dBy_dt, dBz_dt]\n",
+    "if divE_cleaning:\n",
+    "    fields_list.append(dF_dt)\n",
+    "if divB_cleaning:\n",
+    "    fields_list.append(dG_dt)\n",
+    "dEBFG_dt = zeros(dim, 1)\n",
+    "for i in range(dEBFG_dt.shape[0]):\n",
+    "    dEBFG_dt[i] = fields_list[i]\n",
+    "dEBFG_dt = dEBFG_dt.expand()\n",
+    "\n",
+    "# Define matrix M representing the linear system of ODEs\n",
+    "M = zeros(dim)\n",
+    "for i in range(M.shape[0]):\n",
+    "    for j in range(M.shape[1]):\n",
+    "        M[i,j] = dEBFG_dt[i].coeff(EBFG[j], 1)\n",
+    "print(r'M = ')\n",
+    "display(M)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Diagonalization of $W_1 = M (t-t_n) = P_1 D_1 P_1^{-1}$:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "\n",
+    "# Compute matrices of eigenvectors and eigenvalues for diagonalization of W1\n",
+    "W1 = M * (t-tn)\n",
+    "P1, D1 = W1.diagonalize()\n",
+    "invP1 = P1**(-1)\n",
+    "expD1 = exp(D1)\n",
+    "check_diag(W1, D1, P1, invP1)\n",
+    "print('P1 = ')\n",
+    "display(P1)\n",
+    "print('D1 = ')\n",
+    "display(D1)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Diagonalization of $W_2 = M t = P_2 D_2 P_2^{-1}$:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "\n",
+    "# Compute matrices of eigenvectors and eigenvalues for diagonalization of W2\n",
+    "W2 = M * t\n",
+    "P2, D2 = W2.diagonalize()\n",
+    "invP2 = P2**(-1)\n",
+    "expD2 = exp(D2)\n",
+    "check_diag(W2, D2, P2, invP2)\n",
+    "print('P2 = ')\n",
+    "display(P2)\n",
+    "print('D2 = ')\n",
+    "display(D2)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Diagonalization of $W_3 = -M s = P_3 D_3 P_3^{-1}$:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "\n",
+    "# Compute matrices of eigenvectors and eigenvalues for diagonalization of W3\n",
+    "W3 = (-1) * M * s\n",
+    "P3, D3 = W3.diagonalize()\n",
+    "invP3 = P3**(-1)\n",
+    "expD3 = exp(D3)\n",
+    "check_diag(W3, D3, P3, invP3)\n",
+    "print('P3 = ')\n",
+    "display(P3)\n",
+    "print('D3 = ')\n",
+    "display(D3)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### General solution (homogeneous system):\n",
```

Comment:
Could you move this `markdown` cell up, so that it appears before the diagonalization code?
I think it would help the reader understand why we are doing the diagonalization.

Reply:
Yes, thank you. Done in b26e018.